00_Base/src/services/credentials.service.ts

```diff
@@ -474,6 +474,85 @@ export class CredentialsService {
     return updatedClientInformation;
   }
 
+  async deleteTenant(tenantId: string): Promise<void> {
+    const cpoTenant = await this.cpoTenantRepository.readOnlyOneByQuery({
+      where: {
+        id: tenantId,
+      },
+    });
+    if (!cpoTenant) {
+      throw new NotFoundError('CpoTenant not found');
+    }
+
+    const transaction =
+      await this.ocpiSequelizeInstance.sequelize.transaction();
+
+    try {
+      await this.clientInformationRepository.deleteAllByQuery({
+        where: {
+          [ClientInformationProps.cpoTenantId]: cpoTenant.id,
+        },
+      });
+      await this.cpoTenantRepository.deleteAllByQuery(
+        {
+          where: {
+            id: tenantId,
+          },
+        },
+        OcpiNamespace.Credentials,
+      );
+      await transaction.commit();
+      return;
+    } catch (e: any) {
+      await transaction.rollback();
+      throw new InternalServerError(`Could not delete tenant, ${e.message}`);
+    }
+  }
+
+  async unregisterClient(request: UnregisterClientRequestDTO): Promise<void> {
+    const serverPartyId = request.serverPartyId;
+    const serverCountryCode = request.serverCountryCode;
+    const clientPartyId = request.clientPartyId;
+    const clientCountryCode = request.clientCountryCode;
+    const clientInformations =
+      await this.getClientInformationByServerCountryCodeAndPartyId(
+        serverCountryCode,
+        serverPartyId,
+      );
+    if (!clientInformations) {
+      throw new NotFoundError(
+        `Client information not found for server party id ${serverPartyId} and server country code ${serverCountryCode}`,
+      );
+    }
+    const clientInformationMatches = clientInformations.filter(
+      (clientInformation) => {
+        const clientCredntialsRoles =
+          clientInformation[ClientInformationProps.clientCredentialsRoles];
+        return clientCredntialsRoles.some((clientCredntialsRole) => {
+          return (
+            clientCredntialsRole.country_code === clientCountryCode &&
+            clientCredntialsRole.party_id === clientPartyId
+          );
+        });
+      },
+    );
+    if (!clientInformationMatches || clientInformationMatches.length === 0) {
+      throw new NotFoundError(
+        `Client credentials roles not found for client party id ${clientPartyId} and client country code ${clientCountryCode}`,
+      );
+    }
+
+    for (const clientInformation of clientInformationMatches) {
+      const credentialsRoleMatches = clientInformation[ClientInformationProps.clientCredentialsRoles]
+        .filter((clientInformations) => {
+          return clientInformations.country_code === clientCountryCode && clientInformations.party_id === clientPartyId;
+        });
+      for (let credentialsRoleMatch of credentialsRoleMatches) {
+        await credentialsRoleMatch.destroy();
```

Comment:
This removes the credentials on our side, but fails to call DELETE on the client. It is necessary to call DELETE on the client before removing the credentials on our side.

Reply:
Sorry for missing that 🤦 I will add that - also, I presume that for DELETE tenant, we would trigger same behavior for each client tied to that tenant as well, so I will add that in too, but correct me if I am wrong 🙏

Reply:
implemented above